### DISCUSSION TITLE
Is there anyway to improve search performance in high topk sence?

### DISCUSSION THREAD

**Discussion by bingbb47** [milvus-io/milvus#26628] [Q&A and General discussion]
In a sence, the collection has 20M vectors with 152 dim, search on it with Flat index and top30k.
 Is there anyway to improve search performance on this background?

**Comment by bingbb47** [comment]
I have try to do search on a 2.2.14 cluster with (querynodes)6*8c32g, load with 2 replicas.
the serial search latency is about 250 ms.

**Reply by yhmo** [reply]
FLAT index is brute-force search. For each search request, all the 20M vectors are computed. And the top30k is also a big value, I think it makes sense to cost hundreds of milliseconds.

**Reply by xiaofan-luan** [reply]
is there any specific reason k has to be so large?

**Reply by bingbb47** [reply]
Actually, it is a weird requirememt. Although I emphasized its irrationality, still want to make some attempts on Milvus.
I have done some attempt, only the CPU configuration directly affects the latency in this scenario。

but i have tried to create IVF_FLAT index with nlist:256, and search with np:256, it is greater than FLAT. Are these two equivalent？

**Comment by xiaofan-luan** [comment]
yes it's the same.
but usuaully IVF_FLAT nprobe is much smaller than nlist

**Comment by bingbb47** [comment]
I got some new clues.
The target scence is not only has high topk requirement. 
The vector has some weird feature. It was binary vector actually but stored as float vector because they want to cal by IP.
And there are a lot of vectors are duplicate, the differentiation of data is not high, when return 30k items they have same distance.
In the case, ivf index and hnsw index have little positive effect.

I guess that the high latency may caused by sort or reduce step? Do you have any idea?

**Reply by yhmo** [reply]
The time cost is mainly in calculating the distance between vectors. For brute-force search, all the vectors are calculated, no matter whether the vectors are duplicated or not. So, even if you have non-duplicated vectors, the search latency should be similar.
